storage/scratchpad/src/sparse_merkle/mod.rs

```diff
@@ -228,6 +236,349 @@ where
+        
+        // TODO: parallelize up to certain depth.
+        let (sibling_tree, sibling_hashes) =
+            if siblings.len() <= subtree_depth { // Implies sibling_pre_hash is empty.
+                if sibling_updates.is_empty() {
+                    (SubTree::new_empty(), vec![])
+                } else {
+                    Self::batch_create_subtree(
+                        SubTree::new_empty(),
+                        /* target_key = */ sibling_updates[0].0,
+                        /* siblings = */ vec![],
+                        subtree_depth + 1,
+                        sibling_updates,
+                        proof_reader)?
+                }
+            } else {
+                // Only have the sibling hash, need to use proofs.
+                let (subtree, hashes, child_hash) =
+                    Self::batch_create_subtree_by_proof(sibling_updates,
+                                                        proof_reader,
+                                                        sibling_pre_hash,
+                                                        subtree_depth + 1,
+                                                        child_pre_hash)?;
+                child_pre_hash = child_hash;
+                (subtree, hashes)
+            };
+        let (child_tree, child_hashes) = Self::batch_create_subtree(bottom_subtree,
+                                                                    target_key,
+                                                                    siblings,
+                                                                    subtree_depth + 1,
+                                                                    child_updates,
+                                                                    proof_reader)?;
+
+        let (left_tree, right_tree) =
+            Self::swap_if(child_tree, sibling_tree, child_is_right); 
+        let (left_hashes, right_hashes) =
+            Self::swap_if(child_hashes, sibling_hashes, child_is_right);
+        let (left_pre_hash, right_pre_hash) =
+            Self::swap_if(child_pre_hash, sibling_pre_hash, child_is_right);
+        
+        let merged_hashes = Self::merge_txn_hashes(left_pre_hash,
+                                                   left_hashes,
+                                                   right_pre_hash,
+                                                   right_hashes);
+        Ok((SubTree::new_internal(left_tree, right_tree), merged_hashes))
+    }
+
+    /// Given a key and updates, checks if all updates are to this key. If so, generates
+    /// a SubTree for a final leaf, and IntermediateHashes. Each intermediate update is by
+    /// a different transaction as (key, txn_id) pairs are deduplicated.
+    fn leaf_from_updates(
+        leaf_key: HashValue,
+        updates: &[(HashValue, (usize, &V))],
+    ) -> Option<(SubTree<V>, IntermediateHashes)> {
+        if !updates.iter().all(|&u| u.0 == leaf_key) {
```

Comment:
Checking first and last would be fine, since updates are ordered?

Reply:
this is great, thanks!